pkg/registry/core/service/storage/rest_test.go

```diff
@@ -1561,273 +1066,186 @@ func TestServiceRegistryUpdateMultiPortExternalService(t *testing.T) {
 	}
-	if location == nil {
-		t.Errorf("Unexpected nil: %v", location)
-	}
-	if e, a := "https://1.2.3.4:93", location.String(); e != a {
-		t.Errorf("Expected %v, but got %v", e, a)
-	}
-
-	// Test a scheme + name + port (using second ip).
-	location, _, err = redirector.ResourceLocation(ctx, "https:foo-second-ip:p")
+	_, err := storage.Create(ctx,
+		svctest.MakeService("named", svctest.SetPorts(
+			svctest.MakeServicePort("p", 93, intstr.FromInt(80)),
+			svctest.MakeServicePort("q", 76, intstr.FromInt(81)))),
+		rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
 	if err != nil {
-		t.Errorf("Unexpected error: %v", err)
-	}
-	if location == nil {
-		t.Errorf("Unexpected nil: %v", location)
-	}
-	if e, a := "https://[2001:db7::]:93", location.String(); e != a {
-		t.Errorf("Expected %v, but got %v", e, a)
-	}
-
-	// Test a non-existent name + port.
-	_, _, err = redirector.ResourceLocation(ctx, "foo:q")
-	if err == nil {
-		t.Errorf("Unexpected nil error")
-	}
-
-	// Test a non-existent name + port (using second ip).
-	_, _, err = redirector.ResourceLocation(ctx, "foo-second-ip:q")
-	if err == nil {
-		t.Errorf("Unexpected nil error")
-	}
-
-	// Test error path
-	if _, _, err = redirector.ResourceLocation(ctx, "bar"); err == nil {
-		t.Errorf("unexpected nil error")
+		t.Fatalf("unexpected error creating service %q: %v", "named", err)
 	}
+	redirector := rest.Redirector(storage)
 
-	// Test a simple id.
-	_, _, err = redirector.ResourceLocation(ctx, "bad")
-	if err == nil {
-		t.Errorf("Unexpected nil error")
+	cases := []struct {
+		query  string
+		err    bool
+		expect string
+	}{{
+		query:  "unnamed",
+		expect: "//1.2.3.4:80",
+	}, {
+		query:  "unnamed:",
+		expect: "//1.2.3.4:80",
+	}, {
+		query:  "unnamed:93",
+		expect: "//1.2.3.4:80",
+	}, {
+		query:  "http:unnamed:",
+		expect: "http://1.2.3.4:80",
+	}, {
+		query:  "http:unnamed:93",
+		expect: "http://1.2.3.4:80",
+	}, {
+		query: "unnamed:80",
+		err:   true,
+	}, {
+		query:  "unnamed2",
+		expect: "//1.2.3.5:80",
+	}, {
+		query:  "named:p",
+		expect: "//1.2.3.6:80",
+	}, {
+		query:  "named:q",
+		expect: "//1.2.3.6:81",
+	}, {
+		query:  "named:93",
+		expect: "//1.2.3.6:80",
+	}, {
+		query:  "named:76",
+		expect: "//1.2.3.6:81",
+	}, {
+		query:  "http:named:p",
+		expect: "http://1.2.3.6:80",
+	}, {
+		query:  "http:named:q",
+		expect: "http://1.2.3.6:81",
+	}, {
+		query: "named:bad",
+		err:   true,
+	}, {
+		query: "no-endpoints",
+		err:   true,
+	}, {
+		query: "non-existent",
+		err:   true,
+	}}
+	for _, tc := range cases {
+		t.Run(tc.query, func(t *testing.T) {
+			location, _, err := redirector.ResourceLocation(ctx, tc.query)
+			if tc.err == false && err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if tc.err == true && err == nil {
+				t.Fatalf("unexpected success")
+			}
+			if !tc.err {
+				if location == nil {
+					t.Errorf("unexpected location: %v", location)
+				}
+				if e, a := tc.expect, location.String(); e != a {
```

Comment:
same funky if statement. why not just go with `tc.expect != location.String()`

Reply:
it shortens the error line - you don't have to call `location.String()` again